### ISSUE TITLE
Using `as` everywhere in the proxy chain

### ISSUE THREAD

**Issue by cheerfulstoic** [neo4jrb/neo4j#914]
This is an idea from @leviwilson

Right now you can just use `as` at the start of a query chain for models (like `Model.as(:foo)`)

What if we allowed for `as` anywhere in the chain to define node/relationship variables?  Also, what if we took away the arguments for association methods to define those same variables?

I realize that we'd need to do a lot of refactoring.  Also users of the gem would need to as well, so we'd want to deprecate the association variables.  But I think it would make our code simpler and I think it would make the syntax make a lot more sense.

 * There wouldn't be confusion about when to use `as`
 * `optional(:association)` would be cleaner and I think make more sense
 * You could re-define variables when you've been passed a proxy object (I haven't thought this one all the way through)

Would be good to get lots of feedback on this item from lots of people.  Of course feedback from @subvertallchris  would be great (I know you're busy, so no rush.  I don't see this happening anytime soon ;)

**Comment by cheerfulstoic** [comment]
Also I know @jexp was confused by the syntax, so I'd welcome his feedback as well

**Comment by subvertallchris** [comment]
This would be really nice as an option. Things like this would be simpler:

```
def my_method
  some_object.scoped_query.as(:q)
end

private

def self.scoped_query
  all.my_association.where(value: false)
end
```

At the moment, you have to either set the values in your scoped method or give them as arguments. It can be confusing.

I worry about the amount of refactoring it would force for users relative to the benefit offered. Would it be impossible to leave the old version as an option without firing deprecation notices?

**Comment by cheerfulstoic** [comment]
Yeah, certainly we could add the option without the deprecation warnings.  A couple of things to consider, though:

 * I think the deprecations (If we use `ActiveSupport`'s `Deprecation` module) only fire in development and test environments.
 * The deprecation (if implemented correctly) would tell people the lines where it's happening
 * People can fix it by just adding the three characters `.as` in between the association and it's arguments

We could also add a configuration variable so that they could disable the warning if we wanted to.

I think it would simplify our code a lot, though.
